x-pack/plugins/monitoring/public/components/logstash/pipeline_viewer/views/config_viewer/statement_section.js

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import { StatementListHeading } from './statement_list_heading';
+import { Statement } from './statement';
+import { EuiSpacer } from '@elastic/eui';
+
+export function StatementSection({
+  iconType,
+  headingText,
+  elements,
+  onShowVertexDetails
+}) {
+  if (!elements.length) { return null; }
+
+  return (
+    <div className="configStatementList">
+      <StatementListHeading
+        iconType={iconType}
+        title={headingText}
+      />
+      <EuiSpacer size="s" />
+      <StatementList
+        elements={elements}
+        onShowVertexDetails={onShowVertexDetails}
+      />
+    </div>
+  );
+}
+
+function getCollapsedChildIds(elements, collapsedIds) {
+  const collapsedChildIds = new Set();
+  elements.forEach(({ id, parentId }) => {
+    if (collapsedIds.has(parentId) || collapsedChildIds.has(parentId)) {
+      collapsedChildIds.add(id);
+    }
+  });
+  return collapsedChildIds;
+}
+
+class StatementList extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      collapsedIds: new Set(),
+      collapsedChildIds: new Set()
+    };
+  }
+
+  expand = elementId => {
+    const collapsedIds = new Set(this.state.collapsedIds);
+    collapsedIds.delete(elementId);
+    this.updateCollapsedElements(collapsedIds);
+  }
+
+  collapse = elementId => {
+    const collapsedIds = new Set(this.state.collapsedIds);
+    collapsedIds.add(elementId);
+    this.updateCollapsedElements(collapsedIds);
+  }
+
+  updateCollapsedElements = collapsedIds => {
+    const collapsedChildIds = getCollapsedChildIds(this.props.elements, collapsedIds);
+
+    this.setState({
+      collapsedIds,
+      collapsedChildIds
+    });
+  }
+
+  elementIsCollapsed = elementId => this.state.collapsedIds.has(elementId);
+
+  renderStatement = element => {
+    const { id, parentId } = element;
+    const { onShowVertexDetails } = this.props;
+
+    return this.state.collapsedIds.has(parentId) || this.state.collapsedChildIds.has(parentId)
+      ? null
+      : (
+        <Statement
+          key={id}
+          element={element}
+          collapse={this.collapse}
+          expand={this.expand}
+          isCollapsed={this.elementIsCollapsed(id)}
+          onShowVertexDetails={onShowVertexDetails}
+        />
+      );
+  }
+
+  render() {
+    const { elements } = this.props;
+
+    return (
+      <ul className="configViewer__list">
+        {
+          elements.map(this.renderStatement)
+        }
+      </ul>
+    );
+  }
+}
+
+StatementList.propTypes = {
+  elements: PropTypes.arrayOf(
```

Comment:
Nit: WDYT about naming this prop `statements` for parallelism in naming (`StatementList` contains `statements`)?

Reply:
This prop is named `elements` right now because it's expecting a list of instances of the various `*Element` classes, to differentiate from the `Statements` that those objects are composed from.

That said, the component doesn't really care about the class, it cares about the shape of the props. I'm fine with changing it to `statements` for the reason you expressed here.

Reply:
Ah I see, never mind then. If it's expecting a bunch of `*Element` class instances, then naming this prop `elements` is good. Thanks for the clarification.